File: 03-coding-patterns/examples/02-merge-sort/solution.ts
// Merge sort — TypeScript
function mergeSort(arr: number[]): number[] {
    if (arr.length <= 1) return arr.slice();
    const mid = arr.length >> 1;
    return merge(mergeSort(arr.slice(0, mid)), mergeSort(arr.slice(mid)));
}

function merge(left: number[], right: number[]): number[] {
    const out: number[] = [];
    let i = 0, j = 0;
    while (i < left.length && j < right.length) {
        if (left[i]! <= right[j]!) out.push(left[i++]!);
        else out.push(right[j++]!);
    }
    while (i < left.length) out.push(left[i++]!);
    while (j < right.length) out.push(right[j++]!);
    return out;
}

function countInversions(arr: number[]): number {
    let total = 0;
    const buf = new Array<number>(arr.length);

    const sortCount = (a: number[], l: number, r: number): void => {
        if (r - l <= 1) return;
        const m = l + ((r - l) >> 1);
        sortCount(a, l, m); sortCount(a, m, r);
        let i = l, j = m, k = l;
        while (i < m && j < r) {
            if (a[i]! <= a[j]!) buf[k++] = a[i++]!;
            else { buf[k++] = a[j++]!; total += m - i; }
        }
        while (i < m) buf[k++] = a[i++]!;
        while (j < r) buf[k++] = a[j++]!;
        for (let t = l; t < r; ++t) a[t] = buf[t]!;
    };
    sortCount(arr.slice(), 0, arr.length);
    return total;
}

function check(ok: boolean, label: string): void {
    if (!ok) { console.error(`FAIL: ${label}`); process.exit(1); }
}

const arr = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];
const want = [...arr].sort((a, b) => a - b);
check(JSON.stringify(mergeSort(arr)) === JSON.stringify(want), "sort-mixed");
check(JSON.stringify(mergeSort([])) === "[]", "sort-empty");
check(countInversions([2, 4, 1, 3, 5]) === 3, "inversions");

console.log("PASS");

File: 03-coding-patterns/examples/02-merge-sort/solution.cpp
// Merge sort — C++20/23
#include <iostream>
#include <vector>
using namespace std;

static void merge(vector<int>& a, int l, int m, int r, vector<int>& buf) {
    int i = l, j = m, k = l;
    while (i < m && j < r) {
        if (a[i] <= a[j]) buf[k++] = a[i++];
        else buf[k++] = a[j++];
    }
    while (i < m) buf[k++] = a[i++];
    while (j < r) buf[k++] = a[j++];
    for (int t = l; t < r; ++t) a[t] = buf[t];
}

static void merge_sort_rec(vector<int>& a, int l, int r, vector<int>& buf) {
    if (r - l <= 1) return;
    int m = l + (r - l) / 2;
    merge_sort_rec(a, l, m, buf);
    merge_sort_rec(a, m, r, buf);
    merge(a, l, m, r, buf);
}

void merge_sort(vector<int>& a) {
    vector<int> buf(a.size());
    merge_sort_rec(a, 0, static_cast<int>(a.size()), buf);
}

long long count_inversions(vector<int> a) {
    long long total = 0;
    vector<int> buf(a.size());

    auto sort_count = [&](auto& self, int l, int r) -> void {
        if (r - l <= 1) return;
        int m = l + (r - l) / 2;
        self(self, l, m); self(self, m, r);
        int i = l, j = m, k = l;
        while (i < m && j < r) {
            if (a[i] <= a[j]) buf[k++] = a[i++];
            else { buf[k++] = a[j++]; total += (m - i); }
        }
        while (i < m) buf[k++] = a[i++];
        while (j < r) buf[k++] = a[j++];
        for (int t = l; t < r; ++t) a[t] = buf[t];
    };
    sort_count(sort_count, 0, static_cast<int>(a.size()));
    return total;
}

int main() {
    auto check = [](bool ok, const char* label) {
        if (!ok) { cerr << "FAIL: " << label << "\n"; exit(1); }
    };

    {
        vector<int> a{3,1,4,1,5,9,2,6,5,3,5};
        vector<int> want{1,1,2,3,3,4,5,5,5,6,9};
        merge_sort(a);
        check(a == want, "sort-mixed");
    }
    {
        vector<int> a{};
        merge_sort(a);
        check(a.empty(), "sort-empty");
    }
    check(count_inversions({2,4,1,3,5}) == 3, "inversions");

    cout << "PASS\n";
}
